server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java

```diff
@@ -2574,6 +2578,375 @@ public void testDeeplyNestedDocument() throws Exception {
+        assertEquals(1, root.mappers.size());
+        ObjectMapper metrics = (ObjectMapper) root.getMapper("metrics");
+        assertEquals(4, metrics.mappers.size());
+        assertNotNull(metrics.getMapper("service.time"));
+        assertNotNull(metrics.getMapper("service.time.min"));
+        assertNotNull(metrics.getMapper("service.time.max"));
+        assertNotNull(metrics.getMapper("service.time.avg"));
+    }
+
+    public void testSubobjectsFalseArrayMixedContent() throws Exception {
+        DocumentMapper mapper = createDocumentMapper(
+            mapping(b -> b.startObject("metrics").field("type", "object").field("subobjects", false).endObject())
+        );
+        ParsedDocument parsedDocument = mapper.parse(source("""
+            {
+              "metrics.service.time": [
+                1,
+                {
+                  "max" : 2
+                }
+              ]
+            }
+            """));
+
+        assertNotNull(parsedDocument.rootDoc().getField("metrics.service.time"));
+        assertNotNull(parsedDocument.rootDoc().getField("metrics.service.time.max"));
+
+        assertNotNull(parsedDocument.dynamicMappingsUpdate());
+        RootObjectMapper root = parsedDocument.dynamicMappingsUpdate().getRoot();
+        assertEquals(1, root.mappers.size());
+        ObjectMapper metrics = (ObjectMapper) root.getMapper("metrics");
+        assertEquals(2, metrics.mappers.size());
+        assertNotNull(metrics.getMapper("service.time"));
+        assertNotNull(metrics.getMapper("service.time.max"));
+    }
+
+    public void testSubobjectsFalseDocsWithGeoPointFromDynamicTemplate() throws Exception {
+        DocumentMapper mapper = createDocumentMapper(topMapping(b -> {
+            b.startArray("dynamic_templates");
+            {
+                b.startObject();
+                b.startObject("location");
+                {
+                    b.field("match", "location");
+                    b.startObject("mapping");
+                    {
+                        b.field("type", "geo_point");
+                    }
+                    b.endObject();
+                }
+                b.endObject();
+                b.endObject();
+            }
+            b.endArray();
+            b.field("subobjects", false);
+
+        }));
+
+        ParsedDocument parsedDocument = mapper.parse(source("""
+            {
+              "location" : {
+                "lat": 41.12,
+                "lon": -71.34
+              }
+            }
+            """));
+
+        assertNotNull(parsedDocument.rootDoc().getField("location"));
+        RootObjectMapper root = parsedDocument.dynamicMappingsUpdate().getRoot();
+        assertEquals(1, root.mappers.size());
+        Mapper geoPointFieldMapper = root.getMapper("location");
+        assertNotNull(geoPointFieldMapper);
+        assertTrue(geoPointFieldMapper instanceof GeoPointFieldMapper);
+        assertNotNull(parsedDocument.dynamicMappingsUpdate());
+    }
```

Comment:
I think it may be good to add tests that index multiple docs with different way to specify the same fields (flat or not flat). That is because dynamic mappings rely on order and it may happen that we end up only testing the happy path.

Reply:
Added!